### ISSUE TITLE
The app can deadlock

### ISSUE THREAD

**Issue by emilylaguna** [Automattic/pocket-casts-ios#919]
The app can deadlock and freeze up when trying to play episodes. There are multiple ways this can happen and this issue will document them below:

### Expected behavior
The app fails to play the episode but doesn't hand

### Actual behavior
The app hangs

### Steps to reproduce the behavior
1. Add some episodes to your up next that are not downloaded or have been streamed
2. Disable your internet
3. Open your up next queue
4. Tap on an episode in the queue
5. If the Play Up Next on Tap setting is disabled, tap on the Play button in the episode overlay
6. 💥 The app will forever be hanging

### Screenshots
Once the app is hung in Xcode you can click the Pause button to see where it got stuck:

<img width="1361" src="https://github.com/Automattic/pocket-casts-ios/assets/793774/cd711133-c5cf-45b5-9d98-7b92ccae78d7">

**Comment by emilylaguna** [comment]
I did some investigation into this and this is indeed a deadlock happening. 

### What is a deadlock?

> In [concurrent computing](https://en.wikipedia.org/wiki/Concurrent_computing), deadlock is any situation in which no member of some group of entities can proceed because each waits for another member, including itself, to take action, such as sending a message or, more commonly, releasing a [lock](https://en.wikipedia.org/wiki/Lock_(computer_science)).[[1](https://en.wikipedia.org/wiki/Deadlock#cite_note-coulouris-1)]
- https://en.wikipedia.org/wiki/Deadlock


### What does that mean?
In the context of our app the player wants to make sure it state is accurate so it uses a lock to make sure only one area of the app can access it at a time. This is usually done from a background thread, but in this instance a lock is created on a background thread, and then before that is released, another lock is created from the main thread. This results in the background thread having to wait for the main thread, and the main thread having to wait on the background thread. 🔁🔒 

Since the main thread handles all of the UI work and its now stuck waiting for the lock to.. unlock, the app stops responding. 

Here are the actual logs if we add some more info to the lock enter/exit:

```
🔒 Entering Lock. From: load(episode:autoPlay:overrideUpNext:saveCurrentEpisode:completion). Main Thread? No - <NSThread: 0x280332500>{number = 12, name = (null)} - Lock Stack will now be: 1
🔒 Entering Lock. From: load(episode:autoPlay:overrideUpNext:saveCurrentEpisode:completion). Main Thread? No - <NSThread: 0x280332500>{number = 12, name = (null)} - Lock Stack will now be: 2
🔒 Entering Lock. From: cleanupCurrentPlayer. Main Thread? No - <NSThread: 0x280332500>{number = 12, name = (null)} - Lock Stack will now be: 3
🔒 Exiting Lock. From: cleanupCurrentPlayer. Main Thread? No - <NSThread: 0x280332500>{number = 12, name = (null)} - Lock Stack will now be: 2
🔒 Entering Lock. From: cleanupCurrentPlayer. Main Thread? Yes - <_NSMainThread: 0x280c06900>{number = 1, name = main} - Lock Stack will now be: 3

```

And if we look at just the lock stack it looks like:

- Locked on background thread: +1
- Locked on background thread: +2
- Locked on background thread: +3
- Unlocked on background thread: +2
- Locked on main thread: +3
- 💥 🔁 🔒

### What's the fix?
A simple fix could be to just perform the cleanup on a background thread which will prevent the deadlock from happening in this instance, but it doesn't solve the core issue that deadlocks can still happen and this is just one instance of it.

**Comment by joashrajin** [comment]
+1 in 6462353-zd-woothemes

Contains crash logs too

**Comment by rviljoen** [comment]
@emilylaguna I tried to reproduce this, but can no longer do so. I believe this was fixed via https://github.com/Automattic/pocket-casts-ios/pull/1019

**Comment by rviljoen** [comment]
> @emilylaguna I tried to reproduce this, but can no longer do so. I believe this was fixed via #1019

@leandroalonso I think this issue was solved via #1019 and can be closed.

**Comment by leandroalonso** [comment]
Thanks for the ping @rviljoen!
